File: Data Structures & Algorithms/climbing-stairs/submission-1.py
class Solution:
    def climbStairs(self, n: int) -> int:
        
        '''

            f(n) = f(n-1) + f(n-2)
            f(2) = f(1) + f(0) = 1+1= 2

        '''
        if n < 2:
            return 1
        first = 1
        second = 1

        for i in range(2,n+1):
            temp =  second
            second = first + second
            first = temp
        return second

File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-1.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<pair<int,int>> s;
        int maxArea = 0;

        for(int i=0;i<heights.size();i++){
            int start = i;
            while(!s.empty() && s.top().second > heights[i]){
                  pair<int,int> top = s.top();
                  int index = top.first;
                  int height = top.second;
                  s.pop();
                  start = index;
                  maxArea = max(maxArea, (height*(i-index)));
            }
            s.push({start,heights[i]});
           
        }

         while(!s.empty()){
                    int index = s.top().first;
                    int height = s.top().second;
                    
                    maxArea = max(maxArea,(height* (static_cast<int> (heights.size()-index)))); 

                    s.pop();
            }


        return maxArea;
    }
};
